citest/base/execution_context.py

```diff
@@ -0,0 +1,151 @@
+
+  def clear_key(self, key):
+    """Remove key whether or not it exists."""
+    if key in self.__internal:
+      del self.__internal[key]
+    elif key in self.__external:
+      del self.__external[key]
+
+  def get(self, key, default_value):
+    """Lookup value of attribute, or default_value if attribute isn't known."""
+    return (self.__internal[key]
+            if key in self.__internal
+            else self.__external[key] if key in self.__external
+            else default_value)
+
+  def set_snapshotable(self, key, value):
+    """Set attribute key=value, making it appear in snapshots."""
+    if not key:
+      KeyError('Key cannot be empty')
+    if key in self.__internal:
+      KeyError('{0} is already an internal key'.format(key))
+    self.__external[key] = value
+
+  def set_internal(self, key, value):
+    """Set attribute key=value, but not making it appear in snapshots."""
+    if not key:
+      KeyError('Key cannot be empty')
+    if key in self.__external:
+      KeyError('{0} is already a snapshotable key'.format(key))
+    self.__internal[key] = value
+
+  def add_snapshotable(self, key, value):
+    """Set attribute key=value, making it appear in snapshots.
+
+    Raises:
+      KeyError if key is already an attribute.
+    """
+    if not key:
+      KeyError('Key cannot be empty')
+    if key in self.__internal:
+      KeyError('{0} is already an internal key'.format(key))
+    if key in self.__external:
+      KeyError('{0} already exists'.format(key))
+    self.__external[key] = value
+
+  def add_internal(self, key, value):
+    """Set attribute key=value, but not making it appear in snapshots.
+
+    Raises:
+      KeyError if key is already an attribute.
+    """
+    if not key:
+      KeyError('Key cannot be empty')
+    if key in self.__external:
+      KeyError('{0} is already a snapshotable key'.format(key))
+    if key in self.__internal:
+      KeyError('{0} already exists'.format(key))
+    self.__internal[key] = value
+
+  def snapshotable_items(self):
+    """Return list of snapshotable (name, value) tuples."""
+    return self.__external.items()
+
+  def export_to_json_snapshot(self, snapshot, entity):
+    """Implements JsonSnapshotableEntity interface."""
+    for key, value in self.__external.items():
+      snapshot.edge_builder.make_data(entity, key, value)
+
+  def eval(self, value):
+    """Evaluate value in this ExecutionContext.
+
+    Args:
+      value: [any] The value to evaluate.
+
+    Returns:
+      The actual value.
+    """
+    if isinstance(value, list):
+      return [self.eval(elem) for elem in value]
+    elif isinstance(value, dict):
+      return {key: self.eval(data) for key, data in value.items()}
+    elif callable(value):
+      return value(self)
+    else:
+      return value
+
+  def eval_list(self, values):
+    if values is None:
+      return None
+    return [elem(self) if callable(elem) else elem for elem in values]
+
+  def eval_dict(self, values):
```

Comment:
Why have `eval_dict` as well as `eval` which seems to operate on dictionaries, but doesn't call `eval_dict`?.